### PR TITLE
Throw exception when worker cannot handle submission

### PR DESCRIPTION
**Pull request by MTsatsarov**
Related to https://github.com/SoftUni-Internal/OpenJudgeSystem/pull/949

### CODE REVIEW

File: OJS.Workers.SubmissionProcessors/ExecutionTypeFilters/SubmissionFilteringServiceBase.cs [hunk 1]
```diff
@@ -67,8 +68,17 @@ public bool CanProcessSubmission(IOjsSubmission submission, ISubmissionWorker su
...
+                return false;
+            }
+            else
+            {
+                submission.CompilerComment =
+                    "The submission cannot be handled due to unsupported submission type. Please contact an Administrator.";
 
-            return false;
+                throw new Exception($"Submission with Id: {submission.Id}, cannot be processed. Reason: {reason} ");
+            }
```
Comment (gogo4ds) [inline OJS.Workers.SubmissionProcessors/ExecutionTypeFilters/SubmissionFilteringServiceBase.cs:81]: It's not very intuitive to return boolean if it's `true` and `false` only in one single case and throw exception in every other case.
Exception throwing is something we should avoid for better readability, maintainability etc.
Better approach would be to return Enum from this method. e.g `WorkerState` that could be some of the following:
```
Ready,
Unresponsive,
Unhealthy,
DisabledStrategy,
DisabledCompilerType,
```

Based on these the SubmissionProcessor can attach different messages to the submission, without the need to catch exceptions.

**Review by gogo4ds — CHANGES_REQUESTED** [review]

File: OJS.Workers.SubmissionProcessors/ExecutionTypeFilters/SubmissionFilteringServiceBase.cs (hunk 1 above)
Comment (MTsatsarov) [reply OJS.Workers.SubmissionProcessors/ExecutionTypeFilters/SubmissionFilteringServiceBase.cs]: Done.

File: OJS.Workers.SubmissionProcessors/ExecutionTypeFilters/SubmissionFilteringServiceBase.cs [hunk 2]
```diff
@@ -23,11 +23,11 @@ protected SubmissionFilteringServiceBase()
 
         protected abstract ISet<CompilerType> DisabledExecuteAndCompileCompilerTypes { get; }
 
-        public bool CanProcessSubmission(IOjsSubmission submission, ISubmissionWorker submissionWorker)
+        public WorkerStateForSubmission GetWorkerStateForSubmission(IOjsSubmission submission, ISubmissionWorker submissionWorker)
         {
             if (submission == null)
             {
-                return false;
+                return WorkerStateForSubmission.NullableSubmission;
```
Comment (gogo4ds) [inline OJS.Workers.SubmissionProcessors/ExecutionTypeFilters/SubmissionFilteringServiceBase.cs:30]: Submission cannot/should not be null at this point, so you can remove this check.

File: OJS.Workers.SubmissionProcessors/SubmissionProcessor.cs [hunk 3]
```diff
@@ -131,11 +131,52 @@ private IOjsSubmission GetSubmissionForProcessing()
...
+                                $"Worker state for submission: {workerStateForSubmission} is invalid.");
+                    }
+
+                    if (workerStateForSubmission != WorkerStateForSubmission.NullableSubmission)
+                    {
+                        submission.CompilerComment =
+                            "The submission cannot be handled due to unsupported submission type. Please contact an Administrator.";
+                    }
+
+                    this.Logger.Error($"Submission with Id: {submission.Id}, cannot be processed. Reason: {message} ");
```
Comment (gogo4ds) [inline OJS.Workers.SubmissionProcessors/SubmissionProcessor.cs:175]: Make sure this or similar message (with Reason included) is visible by admins in the submission details as well.

**Review by gogo4ds — CHANGES_REQUESTED** [review]

File: OJS.Workers.SubmissionProcessors/ExecutionTypeFilters/SubmissionFilteringServiceBase.cs (hunk 2 above)
Comment (MTsatsarov) [reply OJS.Workers.SubmissionProcessors/ExecutionTypeFilters/SubmissionFilteringServiceBase.cs:30]: Done

File: OJS.Workers.SubmissionProcessors/SubmissionProcessor.cs (hunk 3 above)
Comment (MTsatsarov) [reply OJS.Workers.SubmissionProcessors/SubmissionProcessor.cs:175]: Under the logging, there is a call of OnError method, which will set the message for the admin

File: OJS.Workers.SubmissionProcessors/SubmissionProcessor.cs [hunk 4]
```diff
@@ -129,9 +129,54 @@ private IOjsSubmission GetSubmissionForProcessing()
                     return null;
                 }
 
-                if (!this.submissionsFilteringService.CanProcessSubmission(submission, this.SubmissionWorker))
+                try
```
Comment (gogo4ds) [inline OJS.Workers.SubmissionProcessors/SubmissionProcessor.cs:132]: Why is this in `try`

**Review by gogo4ds — CHANGES_REQUESTED** [review]

**Review by gogo4ds — APPROVED** [review]

**Pull request merged**